Better "unmapped" handling for aggregator factories

Thlamz opened the pull request:
Closes #58139.

Better handling of unmapped cases by providing a default implementation to createUnmapped in ValuesSourceAggregatorFactory and removing any override of createUnmapped that didn't do anything other than instantiate a normal Aggregator with a null ValuesSource. This also allowed me to remove a lot of null handling in the aggregators as nulls are no longer produced. 

I was really conservative in my removal of null checks and overrides of createUnmapped because I didn't want to accidently remove intended behavior. Please tell me if there was some check or override that should also be removed or if I did end up removing something important.

This pull request is not ready to merge yet as some tests in the aggregations modules are failing due to these changes. All these checks simulate unmapped behavior and they are failing because those aggregators are now using the default implementation of createUnmapped, which creates a generic NonCollectingAggregator. I have no idea if the tests should change to reflect these changes or if the default implementation that I provided is insufficient. I opened this PR in this stage because I need feedback to proceed.

Thlamz committed:
Providing default implementation of createUnmapped, removing null checks and using the default implementation of createUnmapped when nothing special was done in the Override

Thlamz committed:
Responding to feedback

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

nik9000 commented:
Oh no! I made a merge conflict for every factory!

Thlamz commented:
@nik9000 Hahaha don't worry, after the PR is reviewed I will rebase and fix the conflitcts, I don't think it should be an issue

nik9000 commented:
> @nik9000 Hahaha don't worry, after the PR is reviewed I will rebase and fix the conflitcts, I don't think it should be an issue

We tend to prefer to make the merge commits in the PR and let github do the squash/rebase stuff.

I think you are waiting on @not-napoleon to review this one, right?

Thlamz commented:
> > @nik9000 Hahaha don't worry, after the PR is reviewed I will rebase and fix the conflitcts, I don't think it should be an issue
> 
> We tend to prefer to make the merge commits in the PR and let github do the squash/rebase stuff.
> 
> I think you are waiting on @not-napoleon to review this one, right?

Thank you, I didn't know that. And yes I am waiting for him but he already responded to me on the issue so I imagine he's taking a look at it soon.

not-napoleon commented on server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregatorFactory.java:
```diff
@@ -52,9 +52,18 @@ public Aggregator createInternal(SearchContext searchContext, Aggregator parent,
+
+            @Override
+            public InternalAggregation buildEmptyAggregation() {
+                return new InternalNonMapped(name, metadata);
```
I suspect this is the source of most of your test failures.  We still need to return the `InternalAggregation` type appropriate to that aggregation class.  I hadn't really thought about that when writing this ticket, and looking at it now I'm not sure we can implement it correctly here on `ValuesSourceAggregationFactory`.  In the spirit of incremental improvement, I suggest we leave `createUnmapped` abstract in `ValuesSourceAggregationFactory` for now, and in each of the aggregations you are changing, we can create an instance of `NonCollectingAggregator` which returns the appropriate `InternalAggregation` type here.

That at least lets us get rid of the `null` handling in the aggregators, which was the main objective here.  Once we have that in place, we can see if a more natural refactoring suggests itself.  I need to give some thought to it.  Long term, it feels like we have too many layers here (Builder -> Factory -> Supplier), and maybe as we work towards cleaning that up, a more natural refactoring will suggest itself.

Thlamz commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/MaxAggregator.java:
```diff
@@ -75,20 +73,12 @@
-            if (parent != null) {
-                return LeafBucketCollector.NO_OP_COLLECTOR;
-            } else {
-                // we have no parent and the values source is empty so we can skip collecting hits.
```
Probably got removed by accident, will fix it with the other feedbacks

Thlamz commented:
Added the assertions you ask for and removed the default implementation of createUnmapped, moving the implementation to each factory where I could use it's InternalAggregator. Had to revert the changes on MissingAggregator and factory because it was failing tests as it seems that there is special behaviour for when there is no valuesSource.

not-napoleon commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/PercentileRanksAggregatorFactory.java:
```diff
@@ -70,7 +70,7 @@ protected Aggregator createUnmapped(SearchContext searchContext,
 
         return percentilesConfig.createPercentileRanksAggregator(name, null, searchContext, parent, percents, keyed,
-                config.format(), metadata);
+            config.format(), metadata);
```
Nit: It's not worth touching this file just for this white space change

not-napoleon commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/MinAggregator.java:
```diff
@@ -76,20 +75,12 @@
-                return LeafBucketCollector.NO_OP_COLLECTOR;
-            } else {
-                // we have no parent and the values source is empty so we can skip collecting hits.
-                throw new CollectionTerminatedException();
```
I think we still want this optimization, which probably means we need to override `getLeafCollector` in `MinAggregatorFactory#createUnmapped()`

not-napoleon commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/StatsAggregatorFactory.java:
```diff
@@ -53,13 +56,19 @@ static void registerAggregators(ValuesSourceRegistry.Builder builder) {
+        };
     }
 
+
```
Nit: unnecessary

not-napoleon reviewed (commented):
This looks a lot better.  You have some formatting-only changes that don't really belong here, and you need to fix the merge conflicts.  After that, ping me again and I'll get it set up in CI.  Thanks for your work!

P.S. If you don't mind my asking, how did you get interested in contributing to Elasticsearch?  These aren't tickets I'd have expected a new contributor to pick up.

Thlamz committed:
Changes based on feedback

Thlamz committed:
Undoing random formatting that happened during changes

Thlamz committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into thlamz-unmaped-handling

 Conflicts:
	server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregatorFactory.java
	server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateHistogramAggregatorFactory.java
	server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregatorFactory.java
	server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregatorFactory.java
	server/src/main/java/org/elasticsearch/search/aggregations/bucket/range/BinaryRangeAggregatorFactory.java
	server/src/main/java/org/elasticsearch/search/aggregations/metrics/AvgAggregatorFactory.java
	server/src/main/java/org/elasticsearch/search/aggregations/metrics/CardinalityAggregatorFactory.java
	server/src/main/java/org/elasticsearch/search/aggregations/metrics/ExtendedStatsAggregatorFactory.java
	server/src/main/java/org/elasticsearch/search/aggregations/metrics/GeoBoundsAggregator.java
	server/src/main/java/org/elasticsearch/search/aggregations/metrics/GeoBoundsAggregatorFactory.java
	server/src/main/java/org/elasticsearch/search/aggregations/metrics/GeoCentroidAggregatorFactory.java
	server/src/main/java/org/elasticsearch/search/aggregations/metrics/MaxAggregatorFactory.java
	server/src/main/java/org/elasticsearch/search/aggregations/metrics/MedianAbsoluteDeviationAggregatorFactory.java
	server/src/main/java/org/elasticsearch/search/aggregations/metrics/MinAggregatorFactory.java
	server/src/main/java/org/elasticsearch/search/aggregations/metrics/StatsAggregatorFactory.java
	server/src/main/java/org/elasticsearch/search/aggregations/metrics/SumAggregatorFactory.java
	server/src/main/java/org/elasticsearch/search/aggregations/metrics/ValueCountAggregatorFactory.java

Thlamz committed:
Fixing imports

Thlamz committed:
Removing yet another unnecessary formatting change

Thlamz committed:
Trying to remove unnecessary formatting

Thlamz commented:
@not-napoleon thanks for the feedback. Had a bit of a fight with IntelliJ who really wanted to remove some random blank spaces from the files I was committing but at last I was able to remove all the unnecessary formatting. I also implemented the getLeafCollector override in MinAggregatorFactory, I had to remove the "final" modifier for this method in NonCollectingAggregator, hope that isn't a problem.

> P.S. If you don't mind my asking, how did you get interested in contributing to Elasticsearch? These aren't tickets I'd have expected a new contributor to pick up.
Contributing was part of a class assignment. I know it is written right there in the Contributing Guidelines that it is not recommended but I did my best to make these changes like I was a contributor invested in this open source project and not like I was just making them to pass my assignment and get this over with. I hope that I put enough effort that my changes actually somewhat mattered to the project.
As to why I chose these tickets specifically: even though I was putting effort to make changes like I was a contributor I knew I wouldn't have enough time to actually learn how everything works in the project so I decided to focus my efforts on a single module or functionality, Analytics/Aggregations in this case. I saw that these issues were open for a while and I understood what changes were needed so I decided to go for it.

not-napoleon commented:
> Had a bit of a fight with IntelliJ who really wanted to remove some random blank spaces from the files I was committing

Yeah, it does that.  It's okay if you're making changes in that file anyway, but if it's the only change in file then it just creates churn in my opinion.  We're also in the process of migrating to a single standard style for the project (as opposed to the somewhat varied personal styles it currently has), so unless you have Intellij configured to use our new style (which there are instructions for doing in the contributing guide) all that formatting is going to get changed again when we roll out spotless.  After loading the new config, I use Intellij's "format changed lines" functionality; if you go to `Code > Reformat File` there's a radial button for "only VCS changed text", which will format just the changes without creating more churn.

>  I hope that I put enough effort that my changes actually somewhat mattered to the project.

For sure.  This is good work, and I appreciate you doing it.

not-napoleon reviewed (approved):
This looks good, and accomplishes the main goal to get rid of all the null checking in the mainline aggregation classes.  As you said earlier, it's a little frustrating to have the duplicated functionality with `buildEmptyAggregation`, I need to do some thinking on if we have a good way to consolidate that.  But just getting rid of the need for null handling is a definite win.

I'll get this into CI and (presuming no issues) merged.  Thanks!

not-napoleon commented:
@elasticmachine, ok to test

Thlamz commented:
@not-napoleon I see that some checks are failing and it looks like it is somewhat related to what I did. I'm having trouble finding out what exactly is wrong by looking at the stack trace. I see that a bunch of tests related to aggregations are failing but I have been unable to reproduce these failures locally. I have been running "gradlew check" and "gradlew :server:test --tests 'org.elasticsearch.search.aggregations.*'" to test my changes and all tests are passing.

Do you see something standing out in the output of the failed tasks that I should take a look at?

not-napoleon commented:
@Thlamz Yeah, these look related.  I don't have a lot of time to dig into it, but I did reproduce one of the failures locally.  It's happening in an integration test, and you can run (that specific example) with: 
```
./gradlew ':rest-api-spec:yamlRestTest' --tests "org.elasticsearch.test.rest.ClientYamlTestSuiteIT" -Dtests.method="test {p0=search.aggregation/10_histogram/date_histogram on range with offset}" -Dtests.seed=41B13CF5281FA29D -Dtests.security.manager=true -Dtests.locale=th-TH -Dtests.timezone=America/Havana -Druntime.java=11
```

I'm not sure why your change would be causing that failure, but since this is a refactoring I would expect it to be invisible to the integration tests.  I realize the stack trace isn't being terribly helpful, but maybe you can turn something up.  Otherwise I might have some time to look into it after the holidays.

Thlamz committed:
Fixing single source of failure

Thlamz commented:
@not-napoleon I believe I was able to fix the specific failed test you were able to reproduce locally. It seems that just replicating the behaviour of BuildEmptyAggregations is not a failproof technique and I believe that most of the other errors are also related to this. I will review my changes and these outputs and try to fix any failed tests. Happy holidays!

Thlamz committed:
More test fixes

Thlamz committed:
Fixing createUnmapped on BinaryRangeAggregatorFactory

Thlamz committed:
Fixing checkstyle
